lib/icinga/notification.cpp

```diff
@@ -428,9 +428,14 @@ void Notification::BeginExecuteNotification(NotificationType type, const CheckRe
 			continue;
 		}
 
+		// Verify if the 'Problem' filter is configured at both the User and Notification object levels.
+		bool foundProblemFilter = NotificationProblem & user->GetTypeFilter() & GetTypeFilter();
+
 		/* on recovery, check if user was notified before */
 		if (type == NotificationRecovery) {
-			if (!notifiedProblemUsers->Contains(userName) && (NotificationProblem & user->GetTypeFilter())) {
+			// Do not send a recovery notification to the current user if he was not previously notified of the
+			// problem state, while containing the 'Problem' filter at both the user and notification object levels.
+			if (!notifiedProblemUsers->Contains(userName) && foundProblemFilter) {
```

Comment:
That comment mostly repeats the condition but doesn't say why. Something like "Don't notify the user about the recovery for a problem they weren't notified about, unless they are explicitly configured to receive recovery notifications but no problem notifications." would be more helpful in my opinion as it gives a bit more context.

Reply:
> unless they are explicitly configured to receive recovery notifications but no problem notifications.

In contrast, this reasoning is not logical to me, as illustrated by the example tests (see PR description), where the user was configured to receive both **recovery** and **problem** states.

Reply:
Okay, maybe it needs a bit more elaboration. What that unless part was supposed to say is that the check only makes sense if it's possible at all for that user to show up in `notifiedProblemUsers` (which is only the case if both the involved `Notification` and `User` object allow problem notifications).